tools/Microsoft.TemplateEngine.Authoring.Tasks/Tasks/ValidateTemplates.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Microsoft.Extensions.Logging;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Authoring.Tasks.Utilities;
+using Microsoft.TemplateEngine.Edge;
+using Microsoft.TemplateEngine.Edge.Settings;
+
+namespace Microsoft.TemplateEngine.Authoring.Tasks.Tasks
+{
+    /// <summary>
+    /// A task that exposes template localization functionality of
+    /// Microsoft.TemplateEngine.TemplateLocalizer through MSBuild.
+    /// </summary>
+    public sealed class ValidateTemplates : Build.Utilities.Task, ICancelableTask
+    {
+        private volatile CancellationTokenSource? _cancellationTokenSource;
+
+        /// <summary>
+        /// Gets or sets the path to the template(s) to be validated.
+        /// </summary>
+        [Required]
+        public string? TemplateLocation { get; set; }
+
+        public override bool Execute()
+        {
+            if (string.IsNullOrWhiteSpace(TemplateLocation))
+            {
+                Log.LogError("The property 'TemplateLocation' should be set for 'ValidateTemplates' target.");
+                return false;
+            }
+
+            string templateLocation = Path.GetFullPath(TemplateLocation);
+
+            using var loggerProvider = new MSBuildLoggerProvider(Log);
+            ILoggerFactory msbuildLoggerFactory = new LoggerFactory(new[] { loggerProvider });
+
+            using CancellationTokenSource cancellationTokenSource = GetOrCreateCancellationTokenSource();
+            CancellationToken cancellationToken = cancellationTokenSource.Token;
+
+            try
+            {
+                using IEngineEnvironmentSettings settings = SetupSettings(msbuildLoggerFactory);
+                Scanner scanner = new(settings);
+                ScanResult scanResult = Task.Run(async () => await scanner.ScanAsync(
+                    templateLocation!,
+                    logValidationResults: false,
+                    returnInvalidTemplates: true,
+                    cancellationToken).ConfigureAwait(false)).GetAwaiter().GetResult();
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                LogResults(scanResult);
+                return !Log.HasLoggedErrors && !cancellationToken.IsCancellationRequested;
```

Comment:
It feels wasteful to fail the result here due to cancellation if we are already done all the processing (the cooperative cancel would not skip any processing)